maps/src/MapBinner.cxx

```diff
@@ -0,0 +1,290 @@
+			out_frame->Put("U",
+			    boost::dynamic_pointer_cast<G3FrameObject>(U_));
+			out_frame->Put("Wpol", map_weights_);
+		} else {
+			out_frame->Put("Wunpol", map_weights_);
+		}
+		out.push_back(out_frame);
+		out.push_back(frame);
+		return;
+	}
+
+	if (frame->type != G3Frame::Scan) {
+		out.push_back(frame);
+		return;
+	}
+
+	if (!boloprops_)
+		log_fatal("Need bolometer properties before detector data "
+		    "can be processed.");
+
+	G3VectorQuatConstPtr pointing =
+	    frame->Get<G3VectorQuat>(pointing_);
+	if (!pointing) {
+		log_error("Missing pointing %s", pointing_.c_str());
+		out.push_back(frame);
+		return;
+	}
+
+	G3TimestreamMapConstPtr timestreams =
+	    frame->Get<G3TimestreamMap>(timestreams_);
+	G3MapDoubleConstPtr weights = frame->Get<G3MapDouble>(weights_);
+	if (!timestreams) {
+		log_error("Missing timestreams %s", timestreams_.c_str());
+		out.push_back(frame);
+		return;
+	}
+	if (!weights) {
+		log_error("Missing weights %s", weights_.c_str());
+		out.push_back(frame);
+		return;
+	}
+
+	if (!units_set_) {
+		T_->units = timestreams->GetUnits();
+		if (Q_)
+			Q_->units = timestreams->GetUnits();
+		if (U_)
+			U_->units = timestreams->GetUnits();
+		units_set_ = true;
+	} else {
+		if (T_->units != timestreams->GetUnits())
+			log_fatal("Timestreams have units that do not match "
+			    "earlier timestreams in the pipeline.");
+	}
+
+#ifdef OPENMP_FOUND
+	// G3SkyMap is not thread-safe when sparse. Avoid the cost of locks
+	// by making one set of sparse maps per thread per scan (which use
+	// essentially zero memory) and then co-adding them at the end of
+	// the scan.
+	G3SkyMapPtr scanT, scanQ, scanU;
+	G3SkyMapWeightsPtr scanW;
+
+	struct chunk {
+		G3SkyMapPtr scanT, scanQ, scanU;
+		G3SkyMapWeightsPtr scanW;
+	};
+	struct chunk *chunk_array;
+	size_t nchunks;
+
+	// Create a list of detectors to satisfy OpenMP's need for scalar
+	// iteration
+	std::vector<std::string> dets;
+	for (auto i : *timestreams)
+		dets.push_back(i.first);
+
+	// Clamp num_threads to prevent memory balloon?
+	#pragma omp parallel private(scanT, scanQ, scanU, scanW) shared(chunk_array,nchunks)
+	{
+		#pragma omp single
+		{
+			chunk_array = new struct chunk[omp_get_num_threads()];
+			nchunks = omp_get_num_threads();
+		}
+
+		#pragma omp barrier
+
+		chunk_array[omp_get_thread_num()].scanT = scanT =
+		    T_->Clone(false);
+		if (Q_)
+			chunk_array[omp_get_thread_num()].scanQ =
+			    scanQ = Q_->Clone(false);
+		if (U_)
+			chunk_array[omp_get_thread_num()].scanU =
+			    scanU = U_->Clone(false);
+		chunk_array[omp_get_thread_num()].scanW = scanW =
+		    map_weights_->Clone(false);
+
+		#pragma omp for
+		for (size_t i = 0; i < dets.size(); i++) {
+			BinTimestream(*timestreams->at(dets[i]),
+			    (weights) ? weights->at(dets[i]) : 1,
+			    boloprops_->at(dets[i]), *pointing,
+			    scanT, scanQ, scanU, scanW);
+		}
+	}
+	for (size_t j = 0; j < nchunks; j++) {
```

Comment:
Might be simpler to just do `T += scanT` in a `pragma omp atomic` block and avoid the chunk array business altogether?

Reply:
`pragma omp atomic` only really works with scalar/POD types, so this would fail. I could move the same chunk_array stuff into a #pragma omp single block, but that seems basically equivalent to the existing code.

Reply:
Sorry, I meant `pragma omp critical` there.

Reply:
Ah, that makes much more sense. That does improve things (change committed).